server/src/main/java/io/druid/query/ResultLevelCachingQueryRunner.java

```diff
@@ -0,0 +1,291 @@
+  }
+
+  @Override
+  public Sequence<T> run(QueryPlus queryPlus, Map responseContext)
+  {
+    if (useResultCache) {
+
+      final String cacheKeyStr = StringUtils.fromUtf8(strategy.computeCacheKey(query));
+      final byte[] cachedResultSet = fetchResultsFromResultLevelCache(cacheKeyStr);
+      String existingResultSetId = extractEtagFromResults(cachedResultSet);
+
+      existingResultSetId = existingResultSetId == null ? "" : existingResultSetId;
+
+      if (query.getContext().get(QueryResource.HEADER_IF_NONE_MATCH) == null) {
+        query = query.withOverriddenContext(
+            ImmutableMap.of(QueryResource.HEADER_IF_NONE_MATCH, existingResultSetId));
+      }
+      Sequence<T> resultFromClient = baseRunner.run(
+          QueryPlus.wrap(query),
+          responseContext
+      );
+      String newResultSetId = (String) responseContext.get(QueryResource.HEADER_ETAG);
+
+      if (newResultSetId != null && newResultSetId.equals(existingResultSetId)) {
+        log.debug("Return cached result set as there is no change in identifiers for query %s ", query.getId());
+        return deserializeResults(cachedResultSet, strategy, existingResultSetId);
+      } else {
+        @Nullable
+        ResultLevelCachePopulator resultLevelCachePopulator = createResultLevelCachePopulator(
+            cacheKeyStr,
+            newResultSetId
+        );
+
+        if (resultLevelCachePopulator == null) {
+          return resultFromClient;
+        }
+        return Sequences.wrap(Sequences.map(
+            resultFromClient,
+            new Function<T, T>()
+            {
+              @Override
+              public T apply(T input)
+              {
+                cacheResultEntry(resultLevelCachePopulator, input);
+                return input;
+              }
+            }
+        ), new SequenceWrapper()
+        {
+          @Override
+          public void after(boolean isDone, Throwable thrown) throws Exception
+          {
+            Preconditions.checkNotNull(resultLevelCachePopulator, "ResultLevelCachePopulator cannot be null during cache population");
+            if (thrown != null) {
+              log.error("Error while preparing for result level caching for query %s ", query.getId());
+            } else {
+              // The resultset identifier and its length is cached along with the resultset
+              resultLevelCachePopulator.populateResults(newResultSetId);
+              log.debug("Cache population complete for query %s", query.getId());
+            }
+          }
+        });
+      }
+    } else {
+      return baseRunner.run(
+          queryPlus,
+          responseContext
+      );
+    }
+  }
+
+  private T cacheResultEntry(
+      ResultLevelCachePopulator resultLevelCachePopulator,
+      T resultEntry
+  )
+  {
+    final Function<T, Object> cacheFn = strategy.prepareForCache(true);
+    resultLevelCachePopulator.cacheObjects
```

Comment:
why do we need to accumulate data chunks inside a list of cacheObjects now and not serialize-and-add them to ByteArrayOutputStream as we are iterating throgh the result set ?

Reply:
@himanshug Cache population has been refactored so that outputstream is used as the results are iterated.